aspnetcore/includes/blazor-security/azuread-roles-groups.md

```diff
@@ -0,0 +1,313 @@
+
+To enable an AAD-registered app for user-defined AAD groups and built-in Administrative Roles, perform the following actions in the Azure portal:
+
+1. Edit the app's manifest (standalone app or Client app of a Hosted solution).
+1. Set `groupMembershipClaims` to `All` to have AAD send a `groups` claim for:
+
+   * AAD Security Groups
+   * Distribution Groups
+   * Administrative Roles
+
+   Set `groupMembershipClaims` to `SecurityGroup` to have AAD send a `roles` claim for:
+
+   * AAD Security Groups
+   * Administrative Roles
+
+1. Assign users to user-defined AAD groups and built-in Administrative Roles. The following examples assume that a user is assigned to the AAD built-in Billing Administrator Role.
+
+The `groups` claim sent by AAD presents the user's groups and roles as Object IDs (GUIDs) in a JSON array. The app must convert the JSON array of groups and roles into individual `groups` claims that the app can build [policies](xref:security/authorization/policies) against. Create a custom user factory in the standalone app or Client app of a Hosted solution:
+
+```csharp
+using System.Security.Claims;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components;
+using Microsoft.AspNetCore.Components.WebAssembly.Authentication;
+using Microsoft.AspNetCore.Components.WebAssembly.Authentication.Internal;
+
+public class CustomUserFactory
+    : AccountClaimsPrincipalFactory<RemoteUserAccount>
+{
+    public CustomUserFactory(NavigationManager navigationManager,
+        IAccessTokenProviderAccessor accessor)
+        : base(accessor)
+    {
+    }
+
+    public async override ValueTask<ClaimsPrincipal> CreateUserAsync(
+        RemoteUserAccount account,
+        RemoteAuthenticationUserOptions options)
+    {
+        var initialUser = await base.CreateUserAsync(account, options);
+
+        if (initialUser.Identity.IsAuthenticated)
+        {
+            SetClaimsFromJsonArray(account, initialUser, "roles", true);
+            SetClaimsFromJsonArray(account, initialUser, "groups", true);
+        }
+
+        return initialUser;
+    }
+
+    private void SetClaimsFromJsonArray(RemoteUserAccount account, 
+        ClaimsPrincipal user, string claimName, bool removeOriginalClaim)
+    {
+        if (account.AdditionalProperties.TryGetValue(claimName, out var rolesObj))
+        {
+            var identity = (ClaimsIdentity)user.Identity;
+            var rolesElement = (JsonElement)rolesObj;
+
+            if (rolesElement.ValueKind == JsonValueKind.Array)
+            {
+                foreach (var element in rolesElement.EnumerateArray())
+                {
+                    identity.AddClaim(new Claim(claimName, element.GetString()));
+                }
+
+                if (removeOriginalClaim)
+                {
+                    identity.TryRemoveClaim(
+                        identity.FindFirst(c => c.Type == claimName));
+                }
+            }
+        }
+    }
```

Comment:
You don't need to do all this manual json parsing. The recommended way to do this is to extend RemoteUserAccount, add a roles or group property with the shape you want (an array in this case) and then make the accountclaimsprincipalfactory be AccountClaimsPrincipalFactory<CustomAccount> and simple do
```
foreach(var role in account.Roles){
  identity.AddClaim(new Claim("role", role);
}
```

The whole point of having an account type is so that you can customize it to do things like this and more clearly map to your domain.

Reply:
> more clearly map to your domain

I have everything in place except for one thing ... I don't understand the client-side mapping step. We did something similar for Identity Server, but this seems different API-wise.

All I see are the `AdditionalProperties` dict in the factory, and I don't think you mean for me to manually do anything with that.

How do I configure with `AddMsalAuthentication`, `AddApiAuthorization`, or `Configure<T>` to automatically map the `roles` and `groups` claim arrays into the extended `RemoteUserAccount`'s `Roles` and `Groups` array properties?